lottie/src/main/java/com/airbnb/lottie/manager/ImageAssetManager.java

```diff
@@ -62,37 +63,54 @@ public void setDelegate(@Nullable ImageAssetDelegate assetDelegate) {
 
   @Nullable public Bitmap bitmapForId(String id) {
     Bitmap bitmap = bitmaps.get(id);
-    if (bitmap == null) {
-      LottieImageAsset imageAsset = imageAssets.get(id);
-      if (imageAsset == null) {
-        return null;
-      }
-      if (delegate != null) {
-        bitmap = delegate.fetchBitmap(imageAsset);
-        if (bitmap != null) {
-          bitmaps.put(id, bitmap);
-        }
-        return bitmap;
+    if (bitmap != null) {
+      return bitmap;
+    }
+
+    LottieImageAsset imageAsset = imageAssets.get(id);
+    if (imageAsset == null) {
+      return null;
+    }
+
+    if (delegate != null) {
+      bitmap = delegate.fetchBitmap(imageAsset);
+      if (bitmap != null) {
+        bitmaps.put(id, bitmap);
       }
+      return bitmap;
+    }
+
+    String filename = imageAsset.getFileName();
+    BitmapFactory.Options opts = new BitmapFactory.Options();
+    opts.inScaled = true;
+    opts.inDensity = 160;
 
-      InputStream is;
+    if (filename.startsWith("data:") && filename.indexOf("base64,") > 0) {
+      // Contents look like a base64 data URI, with the format data:image/png;base64,<data>.
+      byte[] data;
       try {
-        if (TextUtils.isEmpty(imagesFolder)) {
-          throw new IllegalStateException("You must set an images folder before loading an image." +
-              " Set it with LottieComposition#setImagesFolder or LottieDrawable#setImagesFolder");
-        }
-        is = context.getAssets().open(imagesFolder + imageAsset.getFileName());
-      } catch (IOException e) {
-        Log.w(L.TAG, "Unable to open asset.", e);
+        data = Base64.decode(filename.substring(filename.indexOf(',') + 1), Base64.DEFAULT);
```

Comment:
Does the encoding of the base64 string not matter? It looks like that part of the string is never used

Reply:
Yeah, I wondered about possible edge cases here as well, but I believe it should be fine to skip. From my testing, it seems like the ability to resolve a file suffixes (e.g. `.jpg`) to a universally recognized MIME types (e.g. `image/jpeg`) isn't actually required to interpret the binary data of a bitmap resource. This makes some sense, since virtually every bitmap encoding has a [file signature](https://en.wikipedia.org/wiki/List_of_file_signatures). For example, with [PNG](http://www.libpng.org/pub/png/spec/1.2/PNG-Structure.html), the first 8 bytes (hexadecimal) are _always_ `89 50 4E 47 0D 0A 1A 0A`. And everything that knows how to display a PNG basically knows those 8 bytes identify a bitmap image as having PNG encoding, and that the actual image data begins after the file signature.

Of note, the `BitmapFactory.Options` class has an [`outMimeType`](https://developer.android.com/reference/android/graphics/BitmapFactory.Options.html#outMimeType), but no corresponding `inMimeType`. I take that to be Android's way of saying MIME type will always be inferred if we know how to render this image at all.

I also tested what happens when you provide valid base64 data that isn't actually image data, and got a "good" result (no crash) :smile: using `data:image/foobar;base64,YmFy`.